### PR TITLE
[PAY-283][PAY-284] Integrate Tipping Notifications on Web and Mobile

### PR DESCRIPTION
**Pull request by piazzatron**
### Description
- Wires up all things tipping notifs
- The only outstanding work here is to integrate reaction writes (requires a libs change)
Screen Shot 2022-06-03 at 7 20 06 PM
Screen Shot 2022-06-03 at 7 19 59 PM
Screen Shot 2022-06-03 at 7 19 52 PM
Screen Shot 2022-06-03 at 7 19 43 PM
Screen Shot 2022-06-03 at 7 19 33 PM
Screen Shot 2022-06-03 at 7 19 27 PM

### Dragons

*Is there anything the reviewer should be on the lookout for? Are there any dangerous changes?*

### How Has This Been Tested?
Locally, against stage.

*Please describe the tests that you ran to verify your changes. Provide repro instructions & any configuration.*

### How will this change be monitored?

*For features that are critical or could fail silently please describe the monitoring/alerting being added.*

### CODE REVIEW

File: packages/mobile/src/screens/notifications-screen/NotificationBlock.tsx [hunk 1]
```diff
@@ -89,11 +89,11 @@ const typeIconMap: Record<
   [NotificationType.ChallengeReward]: () => IconAudius,
   [NotificationType.TierChange]: (notification: TierChange) =>
     tierInfoMap[notification.tier].icon,
-  [NotificationType.TipReaction]: () => IconTrending,
-  [NotificationType.TipReceived]: () => IconTrending,
-  [NotificationType.TipSent]: () => IconTrending,
-  [NotificationType.TopSupporter]: () => IconTrending,
-  [NotificationType.TopSupporting]: () => IconTrending
+  [NotificationType.Reaction]: () => IconTrending,
+  [NotificationType.TipReceive]: () => IconTrending,
```
Comment (rickyrombo) [inline packages/mobile/src/screens/notifications-screen/NotificationBlock.tsx:93]: Why the tense change :eyes:

File: packages/mobile/src/screens/notifications-screen/NotificationBlock.tsx [hunk 2]
```diff
@@ -182,6 +182,7 @@ export const NotificationBlock = (props: NotificationBlockProps) => {
     state => getNotificationEntities(state, notification),
     isEqual
   )
+  console.log('IN NOTIFICATION BLOCK!!!')
```
Comment (rickyrombo) [inline packages/mobile/src/screens/notifications-screen/NotificationBlock.tsx:185]: log

File: packages/web/src/common/store/notifications/types.ts [hunk 3]
```diff
@@ -174,35 +174,54 @@ export type TierChange = BaseNotification & {
   user: User
 }
 
-export type TipReaction = BaseNotification & {
-  type: NotificationType.TipReaction
+// TODO: when we support multiple reaction types, reactedToEntity type
```
Comment (rickyrombo) [inline packages/web/src/common/store/notifications/types.ts:177]: nit: We could make a `Entity.Tip` type so we're ready?

File: packages/web/src/components/notification/Notification/TipReceivedNotification.tsx [hunk 4]
```diff
@@ -80,11 +124,11 @@ export const TipReceivedNotification = (
           {reactionList.map(([reactionType, Reaction]) => (
             <Reaction
               key={reactionType}
-              onClick={() => setActiveReaction(reactionType)}
+              onClick={() => setReactionValue(reactionType)}
               isActive={
-                activeReaction === null
-                  ? undefined
-                  : reactionType === activeReaction
+                reactionValue // treat 0 and null equivilently here
```
Comment (rickyrombo) [inline packages/web/src/components/notification/Notification/TipReceivedNotification.tsx:129]: super nit: typo equivalently*

**Comment by audius-infra** [comment]
Preview this change https://demo.audius.co/piazz-integrate-tipping-notifs

File: packages/web/src/components/notification/Notification/TipReceivedNotification.tsx [hunk 5]
```diff
@@ -29,18 +40,51 @@ const messages = {
...
+
+  useEffect(() => {
+    const fn = async () => {
+      const reaction = await apiClient.getReaction({
+        reactedToIds: tipTxSignature
+      })
+      if (!reaction) return null
+      setDiscoveryReactionValue(reaction.reactionValue)
+    }
+    fn()
```
Comment (rickyrombo) [inline packages/web/src/components/notification/Notification/TipReceivedNotification.tsx:66]: nit: Maybe only call this if we don't already have a discovery value so it doesn't call discovery on every render? Or some other memoized technique

**Review by rickyrombo — APPROVED** [review]

File: packages/mobile/src/screens/notifications-screen/NotificationBlock.tsx (hunk 2 above)
Comment (piazzatron) [reply packages/mobile/src/screens/notifications-screen/NotificationBlock.tsx:185]: thx

File: packages/mobile/src/screens/notifications-screen/NotificationBlock.tsx (hunk 1 above)
Comment (piazzatron) [reply packages/mobile/src/screens/notifications-screen/NotificationBlock.tsx:93]: Matches the constants from dn -> identity directly this way

File: packages/web/src/components/notification/Notification/TipReceivedNotification.tsx (hunk 4 above)
Comment (piazzatron) [reply packages/web/src/components/notification/Notification/TipReceivedNotification.tsx:129]: thx

File: packages/web/src/common/store/notifications/types.ts (hunk 3 above)
Comment (piazzatron) [reply packages/web/src/common/store/notifications/types.ts:177]: let's cross the bridge when we come to it

File: packages/web/src/components/notification/Notification/TipReceivedNotification.tsx (hunk 5 above)
Comment (piazzatron) [reply packages/web/src/components/notification/Notification/TipReceivedNotification.tsx:66]: discussed via slack

**Comment by audius-infra** [comment]
Preview this change https://demo.audius.co/piazz-integrate-tipping-notifs

**Review by dylanjeffers — APPROVED** [review]
looks awesome btw

**Comment by piazzatron** [comment]
> #### ️✅ There are no secrets present in this pull request anymore.
> 
> If these secrets were true positive and are still valid, we highly recommend you to revoke them. Once a secret has been leaked into a git repository, you should consider it compromised, even if it was deleted immediately. Find [here](https://docs.gitguardian.com/internal-repositories-monitoring/incidents/remediate#2-revoke-the-secret) more information about risks.
> 
> 🦉 [GitGuardian](https://dashboard.gitguardian.com/auth/login/?utm_medium=checkruns&utm_source=github&utm_campaign=cr1) detects secrets in your source code to help developers and security teams secure the modern development process. You are seeing this because you or someone else with access to this repository has authorized GitGuardian to scan your pull request.Our GitHub checks need improvements? [Share your feedbacks](https://050127dayho.typeform.com/to/KmeAPTMk)!

This was not a real secret, BTW. I committed the URL_OVERRIDE in the dev template and it caused GitGuardian to get confused

**Comment by audius-infra** [comment]
Preview this change https://demo.audius.co/piazz-integrate-tipping-notifs

**Comment by audius-infra** [comment]
Preview this change https://demo.audius.co/piazz-integrate-tipping-notifs

File: packages/mobile/src/screens/notifications-screen/Notifications/TopSupporterNotification.tsx [hunk 6]
```diff
@@ -1,39 +1,42 @@
...
   NotificationTile,
   NotificationHeader,
   NotificationTitle,
-  ProfilePicture,
-  UserNameLink,
-  NotificationText
+  NotificationText,
+  ProfilePicture
 } from '../Notification'
+import { UserNameLink } from '../Notification/UserNameLink'
```
Comment (rickyrombo) [inline packages/mobile/src/screens/notifications-screen/Notifications/TopSupporterNotification.tsx:17]: Did this file move?

**Review by rickyrombo — APPROVED** [review]

File: packages/mobile/src/screens/notifications-screen/Notifications/TipReceivedNotification.tsx [hunk 7]
```diff
@@ -29,19 +36,40 @@ const messages = {
   reactionSent: 'Reaction Sent!'
 }
 
+const useReaction = (signature: string) => {
+  return useSelectorWeb(makeGetReactionForSignature(signature))
```
Comment (dylanjeffers) [inline packages/mobile/src/screens/notifications-screen/Notifications/TipReceivedNotification.tsx:40]: i think we want to memoize the makeGetReactionForSignature result so the selector doesn't see a new function coming in? not sure if that's a requirement for useSelector not triggering re-render tho....

**Review by dylanjeffers — APPROVED** [review]
amazing job! can double check all the tipping ui interactions on device and staging

File: packages/mobile/src/screens/notifications-screen/Notifications/TipReceivedNotification.tsx [hunk 8]
```diff
@@ -29,19 +36,40 @@ const messages = {
   reactionSent: 'Reaction Sent!'
 }
 
+const useReaction = (signature: string) => {
```
[the remark this answers is not in the document]
Comment (piazzatron) [reply packages/mobile/src/screens/notifications-screen/Notifications/TipReceivedNotification.tsx:39]: Yeah agreed, unnecessary. Some legacy cruft from when I was trying to make a cool polymorphic hook for desktop/web but couldn't finagle the types. Will remove

File: packages/mobile/src/screens/notifications-screen/Notifications/TipReceivedNotification.tsx (hunk 7 above)
Comment (piazzatron) [reply packages/mobile/src/screens/notifications-screen/Notifications/TipReceivedNotification.tsx:40]: good call

File: packages/web/src/common/store/ui/reactions/sagas.ts [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+  // Add them to the store
+  // Many of these reactions may be null (i.e. entity not reacted to), ignore them
+  const toUpdate = reactions
+    .filter(removeNullable)
+    .map(({ reactedTo, reactionValue }) => ({
+      reaction: getReactionFromRawValue(reactionValue) || 'heart', // this shouldn't happen, default to heart
+      entityId: reactedTo
+    }))
+
+  console.log({ toUpdate })
```
Comment (sddioulde) [inline packages/web/src/common/store/ui/reactions/sagas.ts:36]: log

File: packages/web/src/common/store/ui/reactions/slice.ts [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+
+// The order these reactions appear in the web + mobile UI
+export const reactionOrder: ReactionTypes[] = [
+  'heart',
+  'fire',
+  'party',
+  'explode'
+]
+
+export const reactionsMap: { [k in ReactionTypes]: number } = {
```
Comment (sddioulde) [inline packages/web/src/common/store/ui/reactions/slice.ts:16]: what about sthg like this to make sure the order is always kept?

```
type ReactionTypesMap = { [k in ReactionTypes]: number }
const reactionsMap: ReactionTypesMap = reactionOrder.reduce((acc, curr, i) => ({ ...acc, [curr]: i+1}), {} as ReactionTypesMap)
```

File: packages/web/src/services/audius-api-client/AudiusAPIClient.ts [hunk 11]
```diff
@@ -1378,6 +1392,33 @@ class AudiusAPIClient {
...
+    ) // Perform without retries, using 'split' approach for multiple query params
+
+    if (!response || !response.data.length) return null
+
+    const adapted = response.data.map(item => ({
+      reactionValue: parseInt(item.reaction_value),
+      reactionType: item.reaction_type,
+      senderUserId: decodeHashId(item.sender_user_id),
+      reactedTo: item.reacted_to
+    }))[0]
```
Comment (sddioulde) [inline packages/web/src/services/audius-api-client/AudiusAPIClient.ts:1417]: wondering why we pass an array of ids if we always just read the first one?

**Review by sddioulde — APPROVED** [review]
couple comments, but looks great overall!

File: packages/web/src/common/store/ui/reactions/slice.ts [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+      }>
+    ) => {},
+    fetchReactionValues: (
+      state,
+      action: PayloadAction<{ entityIds: string[] }>
+    ) => {}
+  }
+})
+
+export const makeGetReactionForSignature = (signature: string) => (
```
[the remark this answers is not in the document]
Comment (piazzatron) [reply packages/web/src/common/store/ui/reactions/slice.ts:73]: discussed on slack

File: packages/web/src/common/store/ui/reactions/slice.ts (hunk 10 above)
Comment (piazzatron) [reply packages/web/src/common/store/ui/reactions/slice.ts:16]: Good q - don't think we want to couple the order with the raw value of the reaction here. Imagine we want to change the order on the client someday - we wouldn't want all the old heart reactions to become fire reactions, etc

File: packages/web/src/services/audius-api-client/AudiusAPIClient.ts (hunk 11 above)
Comment (piazzatron) [reply packages/web/src/services/audius-api-client/AudiusAPIClient.ts:1417]: great q - this endpoint is sort of broken rn, in that it accepts multiple tx signatures but only returns the first one. I have a linear tracking fixing this

**Comment by audius-infra** [comment]
Preview this change https://demo.audius.co/piazz-integrate-tipping-notifs

**Comment by audius-infra** [comment]
Preview this change https://demo.audius.co/piazz-integrate-tipping-notifs

**Pull request merged**